server/src/watchdog.js

```diff
@@ -0,0 +1,141 @@
+const config = require("./config").getProperties();
+const mozlog = require("./logging").mozlog("server");
+const db = require("./db");
+const { Shot } = require("./servershot");
+const util = require("util");
+const { URL } = require("url");
+const fetch = require("node-fetch");
+const Hawk = require("hawk");
+const FormData = require("form-data");
+const genUuid = require("nodify-uuid");
+
+const isEnabled = config.watchdog.enable;
+const isProperlyConfigured = validateConfiguration();
+
+function validateConfiguration() {
+  if (!isEnabled) {
+    // Do no warn about config values if Watchdog is not enabled.
+    return false;
+  }
+
+  let isConfigValid = true;
+
+  if (!config.watchdog.id || !config.watchdog.key || !config.watchdog.algorithm) {
+    mozlog.warn("watchdog-invalid-config", {msg: "Watchdog credentials configuration values are missing."});
+    isConfigValid = false;
+  }
+
+  if (!config.watchdog.submissionUrl) {
+    mozlog.warn("watchdog-invalid-config", {msg: "Watchdog submission URL configuration value is missing."});
+    isConfigValid = false;
+  } else {
+    try {
+      new URL(config.watchdog.submissionUrl);
+    } catch (e) {
+      mozlog.warn("watchdog-invalid-config", {msg: "Watchdog submission URL is not a valid URL.", err: e});
+      isConfigValid = false;
+    }
+  }
+
+  return isConfigValid;
+}
+
+let positiveEmailList;
+if (config.watchdog.positiveEmail) {
+  positiveEmailList = config.watchdog.positiveEmail.split(";");
+}
+
+let denominator = config.watchdog.throttleDenominator;
+if (denominator <= 1) {
+  denominator = 1;
+}
+let odometer = 0;
```

Comment:
It would be nice if these were a little clearer. Maybe denominator and odometer themselves could be a class? Like, after looking I can see that denominator means "every {denominator} shots, actually submit", but it's not obvious to me.

And below `shouldSubmitShot()` has side-effects that one wouldn't expect from its name (in that it's maintaining and incrementing a counter).

Reply:
Is there any reason to keep a counter at all? Seems simpler to just do a coin flip each time:

"if Math.random() * N < 1, send to watchdog"

Reply:
A counter allows for a strict rate limit.  We didn't define a rate for this in Screenshots, so either approach is probably ok (with the probability approach being simpler).  It's also meant to be temporary because we do want to send 100% of the shots to Watchdog eventually.